File: go/outputs/markdown.go
package outputs

import (
	"fmt"
	"strings"

	"machinestate/models"
)

// RenderMarkdown outputs the report as Markdown
func RenderMarkdown(report *models.Report) string {
	var sb strings.Builder

	// Header
	sb.WriteString("# MachineState Report\n\n")
	sb.WriteString(fmt.Sprintf("**Generated:** %s  \n", report.Timestamp.Format("2006-01-02 15:04:05")))
	sb.WriteString(fmt.Sprintf("**Hostname:** %s  \n", report.Hostname))
	sb.WriteString(fmt.Sprintf("**OS:** %s  \n", report.OS.Name))
	sb.WriteString(fmt.Sprintf("**Kernel:** %s  \n", report.OS.Kernel))
	sb.WriteString(fmt.Sprintf("**Uptime:** %s  \n", report.System.UptimeHuman))
	if report.System.Timezone != "" {
		sb.WriteString(fmt.Sprintf("**Timezone:** %s  \n", report.System.Timezone))
	}
	if report.System.RebootRequired {
		sb.WriteString("**⚠️ Reboot Required**  \n")
	}
	sb.WriteString("\n")

	// Issues Summary
	sb.WriteString("## Issues Summary\n\n")
	critical, warning, info := countIssues(report.Issues)
	if critical > 0 {
		sb.WriteString(fmt.Sprintf("- **%d** Critical issue(s)\n", critical))
	}
	if warning > 0 {
		sb.WriteString(fmt.Sprintf("- **%d** Warning(s)\n", warning))
	}
	if info > 0 {
		sb.WriteString(fmt.Sprintf("- **%d** Info item(s)\n", info))
	}
	if critical == 0 && warning == 0 && info == 0 {
		sb.WriteString("No issues detected.\n")
	}
	sb.WriteString("\n")

	// System
	sb.WriteString("## System\n\n")
	sb.WriteString("| Metric | Value |\n")
	sb.WriteString("|--------|-------|\n")
	sb.WriteString(fmt.Sprintf("| Load Average | %.2f / %.2f / %.2f (1/5/15 min) |\n",
		report.System.LoadAvg1, report.System.LoadAvg5, report.System.LoadAvg15))
	sb.WriteString(fmt.Sprintf("| CPU Cores | %d |\n", report.System.CPUCores))
	sb.WriteString(fmt.Sprintf("| CPU Usage | %.1f%% |\n", report.System.CPUUsage))
	sb.WriteString(fmt.Sprintf("| Memory | %s / %s (%.1f%%) |\n",
		formatBytes(report.System.MemoryUsed), formatBytes(report.System.MemoryTotal), report.System.MemoryPercent))
	if report.System.SwapTotal > 0 {
		sb.WriteString(fmt.Sprintf("| Swap | %s / %s (%.1f%%) |\n",
			formatBytes(report.System.SwapUsed), formatBytes(report.System.SwapTotal), report.System.SwapPercent))
	}
	sb.WriteString("\n")

	// Disk
	sb.WriteString("## Disk\n\n")
	sb.WriteString("| Mount Point | Size | Used | Free | Usage |\n")
	sb.WriteString("|-------------|------|------|------|-------|\n")
	for _, fs := range report.Disk.Filesystems {
		sb.WriteString(fmt.Sprintf("| %s | %s | %s | %s | %.1f%% |\n",
			fs.MountPoint, formatBytes(fs.Total), formatBytes(fs.Used), formatBytes(fs.Free), fs.UsedPercent))
	}
	sb.WriteString("\n")

	// Network
	sb.WriteString("## Network\n\n")
	sb.WriteString("### Interfaces\n\n")
	sb.WriteString("| Interface | State | IPs |\n")
	sb.WriteString("|-----------|-------|-----|\n")
	for _, iface := range report.Network.Interfaces {
		sb.WriteString(fmt.Sprintf("| %s | %s | %s |\n",
			iface.Name, iface.State, strings.Join(iface.IPs, ", ")))
	}
	sb.WriteString("\n")
	sb.WriteString(fmt.Sprintf("**Internet Connectivity:** %s  \n", boolToStatus(report.Network.Connectivity)))
	sb.WriteString(fmt.Sprintf("**Listening Ports:** %d  \n\n", len(report.Network.ListenPorts)))

	// Packages
	sb.WriteString("## Packages\n\n")
	sb.WriteString(fmt.Sprintf("- **Updates Available:** %d\n", report.Packages.UpdatesAvailable))
	if report.Packages.SecurityUpdates > 0 {
		sb.WriteString(fmt.Sprintf("- **Security Updates:** %d\n", report.Packages.SecurityUpdates))
	}
	if report.Packages.BrokenPackages > 0 {
		sb.WriteString(fmt.Sprintf("- **Broken Packages:** %d\n", report.Packages.BrokenPackages))
	}
	sb.WriteString("\n")

	// Services
	sb.WriteString("## Services\n\n")
	if len(report.Services.FailedUnits) > 0 {
		sb.WriteString("### Failed Services\n\n")
		for _, unit := range report.Services.FailedUnits {
			sb.WriteString(fmt.Sprintf("- %s\n", unit))
		}
		sb.WriteString("\n")
	} else {
		sb.WriteString("No failed services.\n\n")
	}

	sb.WriteString("### Top CPU Processes\n\n")
	sb.WriteString("| Process | CPU | Memory | User |\n")
	sb.WriteString("|---------|-----|--------|------|\n")
	for _, p := range report.Services.TopCPU {
		sb.WriteString(fmt.Sprintf("| %s | %.1f%% | %.1f%% | %s |\n",
			p.Name, p.CPU, p.Memory, p.User))
	}
	sb.WriteString("\n")

	// Security
	sb.WriteString("## Security\n\n")
	sb.WriteString(fmt.Sprintf("- **Firewall:** %s\n", boolToActive(report.Security.FirewallActive)))
	sb.WriteString(fmt.Sprintf("- **SSH:** %s\n", boolToEnabled(report.Security.SSHEnabled)))
	if report.Security.FailedLogins24h > 0 {
		sb.WriteString(fmt.Sprintf("- **Failed Logins (24h):** %d\n", report.Security.FailedLogins24h))
	}
	sb.WriteString("\n")

	// Hardware
	sb.WriteString("## Hardware\n\n")
	if report.Hardware.Battery != nil {
		b := report.Hardware.Battery
		sb.WriteString("### Battery\n\n")
		sb.WriteString(fmt.Sprintf("- **Status:** %s\n", b.Status))
		sb.WriteString(fmt.Sprintf("- **Capacity:** %.0f%%\n", b.Capacity))
		sb.WriteString(fmt.Sprintf("- **Health:** %.1f%%\n", b.Health))
		sb.WriteString(fmt.Sprintf("- **Cycle Count:** %d\n", b.CycleCount))
		sb.WriteString("\n")
	}

	if len(report.Hardware.Temperatures) > 0 {
		sb.WriteString("### Temperatures\n\n")
		sb.WriteString("| Sensor | Temperature |\n")
		sb.WriteString("|--------|-------------|\n")
		for _, t := range report.Hardware.Temperatures {
			sb.WriteString(fmt.Sprintf("| %s | %.1f°C |\n", t.Label, t.Current))
		}
		sb.WriteString("\n")
	}

	if len(report.Hardware.CrashReports) > 0 {
		sb.WriteString(fmt.Sprintf("### Crash Reports\n\n%d crash report(s) in /var/crash\n\n", len(report.Hardware.CrashReports)))
	}

	// Docker
	if report.Docker.Available {
		sb.WriteString("## Docker\n\n")
		if !report.Docker.DaemonRunning {
			sb.WriteString("Docker daemon is not running.\n\n")
		} else {
			sb.WriteString(fmt.Sprintf("- **Running Containers:** %d\n", report.Docker.RunningCount))
			sb.WriteString(fmt.Sprintf("- **Stopped Containers:** %d\n", report.Docker.StoppedCount))
			sb.WriteString(fmt.Sprintf("- **Images:** %d (%s)\n", report.Docker.ImageCount, formatBytes(uint64(report.Docker.TotalImageSize))))
			if report.Docker.DanglingImages > 0 {
				sb.WriteString(fmt.Sprintf("- **Dangling Images:** %s\n", formatBytes(uint64(report.Docker.DanglingImages))))
			}
			sb.WriteString("\n")

			if len(report.Docker.Containers) > 0 {
				sb.WriteString("### Running Containers\n\n")
				sb.WriteString("| Name | Image | Status | CPU | Memory |\n")
				sb.WriteString("|------|-------|--------|-----|--------|\n")
				for _, c := range report.Docker.Containers {
					if c.State == "running" {
						sb.WriteString(fmt.Sprintf("| %s | %s | %s | %.1f%% | %.1f%% |\n",
							c.Name, c.Image, c.Status, c.CPUPercent, c.MemoryPercent))
					}
				}
				sb.WriteString("\n")
			}
		}
	}

	// Snaps
	if report.Snaps.Available && len(report.Snaps.Snaps) > 0 {
		sb.WriteString("## Snaps\n\n")
		sb.WriteString(fmt.Sprintf("- **Installed Snaps:** %d\n", len(report.Snaps.Snaps)))
		sb.WriteString(fmt.Sprintf("- **Total Disk Usage:** %s\n", formatBytes(uint64(report.Snaps.TotalDiskUsage))))
		if report.Snaps.PendingRefreshes > 0 {
			sb.WriteString(fmt.Sprintf("- **Pending Refreshes:** %d\n", report.Snaps.PendingRefreshes))
		}
		sb.WriteString("\n")
	}

	// GPU
	if report.GPU.Available && len(report.GPU.GPUs) > 0 {
		sb.WriteString("## GPU\n\n")
		for _, gpu := range report.GPU.GPUs {
			sb.WriteString(fmt.Sprintf("### %s (%s)\n\n", gpu.Name, gpu.Vendor))
			if gpu.Temperature > 0 {
				sb.WriteString(fmt.Sprintf("- **Temperature:** %d°C\n", gpu.Temperature))
			}
			if gpu.Utilization > 0 {
				sb.WriteString(fmt.Sprintf("- **Utilization:** %d%%\n", gpu.Utilization))
			}
			if gpu.MemoryTotal > 0 {
				sb.WriteString(fmt.Sprintf("- **Memory:** %s / %s\n",
					formatBytes(uint64(gpu.MemoryUsed)), formatBytes(uint64(gpu.MemoryTotal))))
			}
			if gpu.PowerDraw > 0 {
				sb.WriteString(fmt.Sprintf("- **Power Draw:** %.1fW\n", gpu.PowerDraw))
			}
			sb.WriteString("\n")
		}
	}

	// Logs
	if report.Logs.Available {
		sb.WriteString("## Logs (24h)\n\n")
		sb.WriteString(fmt.Sprintf("- **Critical:** %d\n", report.Logs.Stats.CriticalCount))
		sb.WriteString(fmt.Sprintf("- **Errors:** %d\n", report.Logs.Stats.ErrorCount))
		sb.WriteString(fmt.Sprintf("- **Warnings:** %d\n", report.Logs.Stats.WarningCount))
		if report.Logs.Stats.OOMEvents > 0 {
			sb.WriteString(fmt.Sprintf("- **OOM Events:** %d\n", report.Logs.Stats.OOMEvents))
		}
		if report.Logs.Stats.KernelPanics > 0 {
			sb.WriteString(fmt.Sprintf("- **Kernel Panics:** %d\n", report.Logs.Stats.KernelPanics))
		}
		if report.Logs.Stats.Segfaults > 0 {
			sb.WriteString(fmt.Sprintf("- **Segfaults:** %d\n", report.Logs.Stats.Segfaults))
		}
		sb.WriteString("\n")

		if len(report.Logs.Stats.TopErrors) > 0 {
			sb.WriteString("### Top Error Patterns\n\n")
			sb.WriteString("| Count | Pattern |\n")
			sb.WriteString("|-------|--------|\n")
			for _, e := range report.Logs.Stats.TopErrors {
				sb.WriteString(fmt.Sprintf("| %d | %s |\n", e.Count, e.Pattern))
			}
			sb.WriteString("\n")
		}
	}

	// Detailed Issues
	if len(report.Issues) > 0 {
		sb.WriteString("## Detailed Issues\n\n")
		for _, issue := range report.Issues {
			severity := strings.ToUpper(issue.Severity)
			sb.WriteString(fmt.Sprintf("### [%s] %s\n\n", severity, issue.Title))
			sb.WriteString(fmt.Sprintf("**Category:** %s  \n", issue.Category))
			sb.WriteString(fmt.Sprintf("**Description:** %s  \n", issue.Description))
			if issue.Fix != "" {
				sb.WriteString(fmt.Sprintf("**Fix:** %s  \n", issue.Fix))
			}
			sb.WriteString("\n")
		}
	}

	return sb.String()
}

func countIssues(issues []models.Issue) (critical, warning, info int) {
	for _, issue := range issues {
		switch issue.Severity {
		case models.SeverityCritical:
			critical++
		case models.SeverityWarning:
			warning++
		case models.SeverityInfo:
			info++
		}
	}
	return
}

func boolToStatus(b bool) string {
	if b {
		return "Connected"
	}
	return "Disconnected"
}

func boolToActive(b bool) string {
	if b {
		return "Active"
	}
	return "Inactive"
}

func boolToEnabled(b bool) string {
	if b {
		return "Enabled"
	}
	return "Disabled"
}

File: zig/src/outputs/markdown.zig
const std = @import("std");
const report = @import("../models/report.zig");

/// Render report as Markdown
pub fn renderMarkdown(allocator: std.mem.Allocator, r: *const report.Report) ![]u8 {
    var alloc_writer = std.Io.Writer.Allocating.init(allocator);
    errdefer alloc_writer.deinit();
    const w = &alloc_writer.writer;

    // Header
    try w.print("# MachineState Report\n\n", .{});
    try w.print("**Host:** {s}  \n", .{r.hostname});
    try w.print("**Generated:** {s}\n\n", .{r.timestamp});
    try w.writeAll("---\n\n");

    // OS Info
    try w.writeAll("## Operating System\n\n");
    try w.print("| Property | Value |\n", .{});
    try w.print("|----------|-------|\n", .{});
    try w.print("| Name | {s} |\n", .{r.os.name});
    try w.print("| Version | {s} |\n", .{r.os.version});
    try w.print("| Kernel | {s} |\n", .{r.os.kernel});
    try w.print("| Architecture | {s} |\n", .{r.os.architecture});
    try w.writeAll("\n");

    // System
    try w.writeAll("## System\n\n");
    try w.print("| Metric | Value |\n", .{});
    try w.print("|--------|-------|\n", .{});
    try w.print("| Uptime | {s} |\n", .{r.system.uptime_human});
    try w.print("| Timezone | {s} |\n", .{r.system.timezone});
    try w.print("| CPU Cores | {d} |\n", .{r.system.cpu_cores});
    try w.print("| CPU Usage | {d:.1}% |\n", .{r.system.cpu_usage});
    try w.print("| Load Average | {d:.2} / {d:.2} / {d:.2} |\n", .{ r.system.load_avg_1, r.system.load_avg_5, r.system.load_avg_15 });
    try w.print("| Memory | {s} / {s} ({d:.1}%) |\n", .{ formatBytes(r.system.memory_used), formatBytes(r.system.memory_total), r.system.memory_percent });
    if (r.system.swap_total > 0) {
        try w.print("| Swap | {s} / {s} ({d:.1}%) |\n", .{ formatBytes(r.system.swap_used), formatBytes(r.system.swap_total), r.system.swap_percent });
    }
    if (r.system.reboot_required) {
        try w.writeAll("| Reboot Required | Yes |\n");
    }
    try w.writeAll("\n");

    // Disk
    try w.writeAll("## Disk\n\n");
    try w.print("| Mount | Device | Used | Total | Usage |\n", .{});
    try w.print("|-------|--------|------|-------|-------|\n", .{});
    for (r.disk.filesystems) |fs| {
        try w.print("| {s} | {s} | {s} | {s} | {d:.1}% |\n", .{ fs.mount_point, fs.device, formatBytes(fs.used), formatBytes(fs.total), fs.used_percent });
    }
    try w.writeAll("\n");

    // Network
    try w.writeAll("## Network\n\n");
    try w.print("| Interface | State | MAC | RX | TX |\n", .{});
    try w.print("|-----------|-------|-----|----|----|", .{});
    try w.writeAll("\n");
    for (r.network.interfaces) |iface| {
        try w.print("| {s} | {s} | {s} | {s} | {s} |\n", .{ iface.name, iface.state, iface.mac, formatBytes(iface.rx_bytes), formatBytes(iface.tx_bytes) });
    }
    try w.writeAll("\n");

    try w.print("**Internet Connectivity:** {s}\n\n", .{if (r.network.connectivity) "Connected" else "Not connected"});

    if (r.network.listen_ports.len > 0) {
        try w.writeAll("**Listening Ports:** ");
        var first = true;
        for (r.network.listen_ports) |port| {
            if (!first) try w.writeAll(", ");
            try w.print("{d}/{s}", .{ port.port, port.protocol });
            first = false;
        }
        try w.writeAll("\n\n");
    }

    // Hardware
    if (r.hardware.battery != null or (r.hardware.temperatures != null and r.hardware.temperatures.?.len > 0)) {
        try w.writeAll("## Hardware\n\n");

        if (r.hardware.battery) |bat| {
            try w.writeAll("### Battery\n\n");
            try w.print("| Property | Value |\n", .{});
            try w.print("|----------|-------|\n", .{});
            try w.print("| Status | {s} |\n", .{bat.status});
            try w.print("| Capacity | {d:.0}% |\n", .{bat.capacity});
            try w.print("| Health | {d:.1}% |\n", .{bat.health});
            try w.print("| Cycles | {d} |\n", .{bat.cycle_count});
            try w.writeAll("\n");
        }

        if (r.hardware.temperatures) |temps| {
            if (temps.len > 0) {
                try w.writeAll("### Temperatures\n\n");
                try w.print("| Sensor | Temperature |\n", .{});
                try w.print("|--------|-------------|\n", .{});
                for (temps) |temp| {
                    if (temp.current > 0) {
                        try w.print("| {s} | {d:.0}C |\n", .{ temp.label, temp.current });
                    }
                }
                try w.writeAll("\n");
            }
        }
    }

    // Packages
    if (r.packages.updates_available > 0 or r.packages.security_updates > 0) {
        try w.writeAll("## Packages\n\n");
        try w.print("| Metric | Value |\n", .{});
        try w.print("|--------|-------|\n", .{});
        try w.print("| Updates Available | {d} |\n", .{r.packages.updates_available});
        if (r.packages.security_updates > 0) {
            try w.print("| Security Updates | {d} |\n", .{r.packages.security_updates});
        }
        try w.writeAll("\n");
    }

    // Services
    if (r.services.zombie_count > 0 or r.services.failed_units.len > 0) {
        try w.writeAll("## Services\n\n");
        if (r.services.zombie_count > 0) {
            try w.print("**Zombie Processes:** {d}\n\n", .{r.services.zombie_count});
        }
        if (r.services.failed_units.len > 0) {
            try w.writeAll("**Failed Units:**\n");
            for (r.services.failed_units) |unit| {
                try w.print("- {s}\n", .{unit});
            }
            try w.writeAll("\n");
        }
    }

    // Docker
    if (r.docker) |docker| {
        if (docker.available) {
            try w.writeAll("## Docker\n\n");
            try w.print("| Metric | Value |\n", .{});
            try w.print("|--------|-------|\n", .{});
            try w.print("| Daemon | {s} |\n", .{if (docker.daemon_running) "Running" else "Stopped"});
            try w.print("| Running Containers | {d} |\n", .{docker.running_count});
            try w.print("| Stopped Containers | {d} |\n", .{docker.stopped_count});
            try w.print("| Images | {d} |\n", .{docker.image_count});
            try w.print("| Total Image Size | {s} |\n", .{formatBytes(@intCast(docker.total_image_size))});
            try w.writeAll("\n");
        }
    }

    // Snaps
    if (r.snaps) |snaps| {
        if (snaps.available) {
            try w.writeAll("## Snaps\n\n");
            try w.print("| Metric | Value |\n", .{});
            try w.print("|--------|-------|\n", .{});
            if (snaps.snaps) |snap_list| {
                try w.print("| Installed | {d} |\n", .{snap_list.len});
            }
            try w.print("| Disk Usage | {s} |\n", .{formatBytes(@intCast(snaps.total_disk_usage))});
            if (snaps.pending_refreshes > 0) {
                try w.print("| Pending Refreshes | {d} |\n", .{snaps.pending_refreshes});
            }
            try w.writeAll("\n");
        }
    }

    // GPU
    if (r.gpu) |gpu| {
        if (gpu.available) {
            if (gpu.gpus) |gpus| {
                try w.writeAll("## GPU\n\n");
                for (gpus) |g| {
                    try w.print("### {s}\n\n", .{g.name});
                    try w.print("| Metric | Value |\n", .{});
                    try w.print("|--------|-------|\n", .{});
                    try w.print("| Vendor | {s} |\n", .{g.vendor});
                    if (g.temperature > 0) {
                        try w.print("| Temperature | {d}C |\n", .{g.temperature});
                    }
                    if (g.utilization > 0) {
                        try w.print("| Utilization | {d}% |\n", .{g.utilization});
                    }
                    if (g.memory_total > 0) {
                        try w.print("| Memory | {s} / {s} |\n", .{ formatBytes(@intCast(g.memory_used)), formatBytes(@intCast(g.memory_total)) });
                    }
                    try w.writeAll("\n");
                }
            }
        }
    }

    // Logs
    if (r.logs) |logs| {
        if (logs.available) {
            try w.writeAll("## Logs (24h)\n\n");
            try w.print("| Metric | Value |\n", .{});
            try w.print("|--------|-------|\n", .{});
            try w.print("| Errors | {d} |\n", .{logs.stats.error_count});
            if (logs.stats.critical_count > 0) {
                try w.print("| Critical | {d} |\n", .{logs.stats.critical_count});
            }
            if (logs.stats.oom_events > 0) {
                try w.print("| OOM Events | {d} |\n", .{logs.stats.oom_events});
            }
            if (logs.stats.kernel_panics > 0) {
                try w.print("| Kernel Panics | {d} |\n", .{logs.stats.kernel_panics});
            }
            if (logs.stats.segfaults > 0) {
                try w.print("| Segfaults | {d} |\n", .{logs.stats.segfaults});
            }
            try w.writeAll("\n");
        }
    }

    // Issues
    if (r.issues.len > 0) {
        try w.writeAll("## Issues\n\n");

        // Count by severity
        var critical: usize = 0;
        var warnings: usize = 0;
        for (r.issues) |issue| {
            if (std.mem.eql(u8, issue.severity, "critical")) {
                critical += 1;
            } else if (std.mem.eql(u8, issue.severity, "warning")) {
                warnings += 1;
            }
        }

        if (critical > 0 or warnings > 0) {
            try w.writeAll("**Summary:** ");
            if (critical > 0) try w.print("{d} critical ", .{critical});
            if (warnings > 0) try w.print("{d} warnings", .{warnings});
            try w.writeAll("\n\n");
        }

        for (r.issues) |issue| {
            const icon = if (std.mem.eql(u8, issue.severity, "critical"))
                "X"
            else if (std.mem.eql(u8, issue.severity, "warning"))
                "!"
            else
                "i";

            try w.print("### [{s}] [{s}] {s}\n\n", .{ icon, issue.category, issue.title });
            try w.print("{s}\n\n", .{issue.description});
            if (issue.fix) |fix| {
                try w.print("**Fix:** {s}\n\n", .{fix});
            }
        }
    }

    // Footer
    try w.writeAll("---\n\n");
    try w.writeAll("*Generated by machinestate*\n");

    return alloc_writer.toOwnedSlice();
}

fn formatBytes(bytes: u64) [16]u8 {
    var buf: [16]u8 = std.mem.zeroes([16]u8);
    const units = [_][]const u8{ "B", "KB", "MB", "GB", "TB" };
    var value: f64 = @floatFromInt(bytes);
    var unit_idx: usize = 0;

    while (value >= 1024 and unit_idx < units.len - 1) {
        value /= 1024;
        unit_idx += 1;
    }

    _ = std.fmt.bufPrint(&buf, "{d:.1} {s}", .{ value, units[unit_idx] }) catch {};
    return buf;
}
